pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableFSTIndex.java

```
@@ -0,0 +1,80 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.realtime.impl.invertedindex;
+
+import java.io.IOException;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableFST;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableFSTImpl;
+import org.apache.pinot.segment.local.utils.nativefst.utils.RealTimeRegexpMatcher;
+import org.apache.pinot.segment.spi.index.mutable.MutableTextIndex;
+import org.roaringbitmap.RoaringBitmapWriter;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+public class NativeMutableFSTIndex implements MutableTextIndex {
+  private final String _column;
+  private final MutableFST _fst;
+  private int _dictId;
+  private final ReentrantReadWriteLock.ReadLock _readLock;
+  private final ReentrantReadWriteLock.WriteLock _writeLock;
+
+  public NativeMutableFSTIndex(String column) {
+    _column = column;
+    _fst = new MutableFSTImpl();
+
+    ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    _readLock = readWriteLock.readLock();
+    _writeLock = readWriteLock.writeLock();
+  }
+
+  @Override
+  public void add(String document) {
+    _writeLock.lock();
+    try {
+      _fst.addPath(document, _dictId);
+      _dictId++;
+    } finally {
+      _writeLock.unlock();
+    }
+  }
+
+  @Override
+  public ImmutableRoaringBitmap getDictIds(String searchQuery) {
+    RoaringBitmapWriter<MutableRoaringBitmap> writer = RoaringBitmapWriter.bufferWriter().get();
+    _readLock.lock();
+    try {
+      RealTimeRegexpMatcher.regexMatch(searchQuery, _fst, writer::add);
+      return writer.get();
+    } finally {
+      _readLock.unlock();
+    }
+  }
+
+  @Override
+  public MutableRoaringBitmap getDocIds(String searchQuery) {
+    return null;
```

Let's throw an unsupported exception with some error message for debugging purpose